tests/federation_room_join_partial_state_test.go

```diff
@@ -933,44 +1088,35 @@ func TestPartialStateJoin(t *testing.T) {
 		// wait for the state_ids request to arrive
 		psjResult.AwaitStateIdsRequest(t)
 
+		// Eager sync shouldn't include the room yet
+		response, nextBatch := alice.MustSync(t, client.SyncReq{})
+
+		syncJoinedRoomPath := "rooms.join." + client.GjsonEscape(serverRoom.RoomID)
+		if response.Get(syncJoinedRoomPath).Exists() {
+			t.Fatal("Sync shouldn't include the joined room until resync is over")
+		}
+
 		// restart the homeserver
 		err := deployment.Restart(t)
 		if err != nil {
 			t.Errorf("Failed to restart homeserver: %s", err)
 		}
 
-		// attempts to sync should block. Fire off a goroutine to try it.
-		syncResponseChan := make(chan gjson.Result)
-		go func() {
-			response, _ := alice.MustSync(t, client.SyncReq{})
-			syncResponseChan <- response
-			close(syncResponseChan)
-		}()
+		// Sync still shouldn't include the room
+		response, nextBatch = alice.MustSync(t, client.SyncReq{Since: nextBatch})
 
-		// we expect another state_ids request to arrive.
-		// we'd do another AwaitStateIdsRequest, except it's single-use.
-
-		// the client-side requests should still be waiting
-		select {
-		case <-syncResponseChan:
-			t.Fatalf("Sync completed before state resync complete")
-		default:
+		if response.Get(syncJoinedRoomPath).Exists() {
+			t.Fatal("Sync shouldn't include the joined room until resync is over")
 		}
 
 		// release the federation /state response
 		psjResult.FinishStateRequest()
 
 		// the /sync request should now complete, with the new room
-		var syncRes gjson.Result
-		select {
-		case <-time.After(1 * time.Second):
-			t.Fatalf("/sync request request did not complete")
-		case syncRes = <-syncResponseChan:
-		}
+		response, _ = alice.MustSync(t, client.SyncReq{Since: nextBatch})
```

Comment:
Now that eager /syncs no longer block, shouldn't we use a `MustSyncUntil` or `awaitPartialStateJoinCompletion`, to avoid the race where the /sync response comes before finishing the partial state join?

Reply:
8301105

Reply:
err, actually 76d2e5c